### PR TITLE
Unified renderer

### LINKED ISSUE #111: Build `wgpu` renderer
`wgpu` is a crate that supports multiple [backends](https://docs.rs/wgpu/latest/wgpu/enum.Backend.html).

This would currently leave out DirectX 9, but I wonder if we could leverage it to merge the DirectX11, DirectX12, OpenGL renderers and also add instant support for Vulkan.

### LINKED ISSUE #142: Incompatibility Issue with Sea of Thieves (DX11)
I am currently experiencing issues while trying to implement the HudHook library into the Sea of Thieves game. Despite following all the instructions provided in the documentation, I encountered several problems that are preventing me from using it smoothly.

Firstly, when utilizing the master branch of the HudHook library, I experience random freezing of the game and I get "Could not lock in WndProc" every ~10sec in the trace console. Occasionally, the game runs without any apparent issues; however, the in-game menu fails to show up. Moreover, I encounter rendering glitches similar to those demonstrated in this video: [Rendering Glitches Video](https://www.youtube.com/watch?v=ZR8cGKQfjVQ)

Secondly, as an alternative approach, I attempted to use the unified-renderer branch instead. While this version resolves the previous rendering concerns, there is now another issue where neither the mouse nor the keyboard functions properly within the ImGUI window. Consequently, controlling or interacting with the elements inside the ImGUI window becomes impossible. To provide further clarification regarding these challenges, please refer to this additional video demonstration: [Demonstration Video](https://www.youtube.com/watch?v=TafFp3wb3qA)

Would you happen to have any suggestions for addressing this issue?I have endeavored to modify your lib to try to fix this issue but without success. 

Despite these hurdles, I genuinely appreciate the efforts invested in developing such a fantastic lib.

Thank you in advance for your time and support.

### PR DESCRIPTION
**Pull request by veeenu**
This is a **MASSIVE** refactor that completely changes the underlying rendering engine, which is now a unified DirectX12 renderer that is then composited over games' windows via [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/directcomposition-portal), with barely any public API change.

It seamlessly works with Dark Souls III and Elden Ring already, and the integration test cases.

Being such a breaking change, I will shamelessly ping all people that may be affected as I don't have a lot of control over corner cases but really don't want to merge this in before I'm sure I'm not disrupting anything. I would be super grateful if you'd be up for checking this out. I expect some of you may have moved on to other projects, in that case I hope you won't be annoyed by the ping!

## Explanation

This PR implements a DirectX 12 renderer in `hudhook::renderer::dx12` which renders over an off-screen surface, and then uses DirectComposition to composite the off-screen surface on top of the HWND, completely sidestepping the games' native rendering engine which will run on its own.

`Present` hooks now only serve the purpose of actually triggering a render and a composition and the device objects aren't touched at all. In fact, most of the hooks' rendering logic except building the actual hooked function implementations is now shared and identical across all hooks.

As a result, hooks are now tiny compared to what they were before and their maintenance is going to be very easy. Similarly, with a single renderer there is much less code to maintain.

Additionally, this will make it _extremely_ simple to add new renderers 

## Affected issues

As a bonus, this PR _should_ fix the following issues, or at the very least give us a cleaner path to their resolution.

If you are getting pinged, it would mean the world if you could give this branch a spin and report back if it fixed the associated issue.

- #79
  @Godnoken @camas since the renderer does not lock DirectX9 specific resources, it could alleviate/solve the issue. (need to investigate)
- #94
  @Godnoken I have found a different way of blocking input that doesn't involve rerouting many input functions: just use a transparent overlay window. It's not implemented but I think we could explore that avenue and it could work for all hooks.
- #103 
  @Godnoken this branch uses `WM_SIZE` for resize events which should sidestep everything that causes your issue altogether. (might close the issue)
- #107 
  @Earthnuker @SkyLeite this greatly reduces the issue surface; now that the hook code is drastically simpler, we could try a number of solutions, provided this one somehow doesn't solve the issue already. (need to investigate)
- Closes #111.
  We won't need a renderer with multiple backends after all!
- #137 
  @vars1ty not sure about the specific game, but maybe things may have changed enough that this now works correctly. (need to investigate)
- Closes #142 
  @Thoxy67 this branch should most definitely work for you now.

Thank you all for your patience and sorry again for the ping! 😅

### CODE REVIEW

File: src/hooks/opengl3.rs [hunk 1]
```diff
@@ -1,246 +1,44 @@
...
-            let mut imgui_renderer = ImguiRenderer {
-                ctx: context,
-                renderer,
-                wnd_proc,
-                game_hwnd: hwnd,
-                resolution_and_rect: None,
-            };
-
-            // Initialize window events on the imgui renderer
-            ImguiWindowsEventHandler::setup_io(&mut imgui_renderer);
```
Comment (veeenu) [inline src/hooks/opengl3.rs:73]: Note to self: this part of logic is missing and should be added

**Comment by camas** [comment]
Unfortunately #79 still an issue. Tested with 7791c49

Fails soon after injection when it can't lock the render engine, with a bunch of WndProc lock fails before.

```
19:00:12 [DEBUG] (1) hudhook::hooks: WNDCLASSEXW { cbSize: 80, style: WNDCLASS_STYLES(3), cbClsExtra: 0, cbWndExtra: 0, hInstance: HINSTANCE(140696328601600), hIcon: HICON(0), hCursor: HCURSOR(0), hbrBackground: HBRUSH(0), lpszMenuName: PCWSTR(0x0), lpszClassName: PCWSTR(0x7ffbf3eff8e8), hIconSm: HICON(0) }
19:00:12 [DEBUG] (1) hudhook::hooks: HWND(788424)
19:00:12 [INFO] Output window: IDXGISwapChain(IUnknown(0x22b60745720))
19:00:12 [INFO] Desc: DXGI_SWAP_CHAIN_DESC { BufferDesc: DXGI_MODE_DESC { Width: 1920, Height: 1080, RefreshRate: DXGI_RATIONAL { Numerator: 0, Denominator: 1 }, Format: DXGI_FORMAT(28), ScanlineOrdering: DXGI_MODE_SCANLINE_ORDER(0), Scaling: DXGI_MODE_SCALING(0) }, SampleDesc: DXGI_SAMPLE_DESC { Count: 1, Quality: 0 }, BufferUsage: DXGI_USAGE(48), BufferCount: 2, OutputWindow: HWND(1771696), Windowed: BOOL(1), SwapEffect: DXGI_SWAP_EFFECT(3), Flags: 2114 }
19:00:25 [DEBUG] (2) hudhook::hooks::render: Could not lock in WndProc
...
19:01:01 [DEBUG] (2) hudhook::hooks::render: Could not lock in WndProc
19:01:01 [ERROR] Could not lock render engine
```

**Comment by veeenu** [comment]
@camas I stalked your GitHub and noticed you are using `hudhook` for Golf With Your Friends. I have that game so I can take a look at it directly! But yeah I can easily see how that's a synchronization issue. It could be solved by using threads and channels but I'm not sure I want the added overhead of those.

Weirdly enough, there is only 3 places now where the render engine is locked that could create contention:
- [here during `render`](https://github.com/veeenu/hudhook/blob/51068020db8e9304be73de171678478cdbbda4b7/src/hooks/render.rs#L77)
- [here during wnd proc](https://github.com/veeenu/hudhook/blob/51068020db8e9304be73de171678478cdbbda4b7/src/hooks/render.rs#L124)
- [here during `resize`](https://github.com/veeenu/hudhook/blob/51068020db8e9304be73de171678478cdbbda4b7/src/hooks/render.rs#L93)

My hunch there is that the wnd proc calls itself recursively, locks once and is not ever able to exit and do anything else. Or the `WND_PROC` static is initialized twice (which shouldn't happen as we use `get_or_init` that has the guarantee of uniqueness) and thus called recursively via `CallWindowProcW`.

**Comment by veeenu** [comment]
I have gotten some feedback and will now merge this in as there are no apparent regressions for the time being.

**Comment by vars1ty** [comment]
Thanks for the ping, feel free to ping me whenever as I'm more than happy to test out features whenever I can.

I'll update to the latest commit, fix a few errors due to the changes and then test it in both OpenGL and then later DirectX9, will let you know how it goes 👍🏽

**Comment by veeenu** [comment]
Thank you 🙏
